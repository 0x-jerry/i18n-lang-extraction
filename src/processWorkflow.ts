import {
  quoteStrRegexp,
  singleChineseRegexp,
  templateStrRegexp,
  tplInterpolationRegexp,
  buildTagMatchRegexp,
  parseStringTemplate,
  processWithoutComments,
  maybeChineseStrRegexp,
} from './utils'

interface ProcessContext extends ProcessOptions {
  content: string
}

export enum ProcessContentType {
  Vue = 'vue',
  Js = 'js',
}

export interface ProcessOptions {
  type?: ProcessContentType | `${ProcessContentType}`
  getI18nKey?(str: string): string | undefined
}

export async function processI18nWorkflow(fileContent: string, opt: ProcessOptions) {
  const ctx: ProcessContext = {
    content: fileContent,
    ...opt,
  }

  return processWithoutComments(fileContent, (contentWithoutComment) => {
    if (ctx.type === 'vue') {
      return processVueContent(ctx, contentWithoutComment)
    } else {
      return processJsContent(ctx, contentWithoutComment)
    }
  })
}

// --------- replacer
function processVueContent(ctx: ProcessContext, fileContent: string) {
  const result = fileContent
    .replace(buildTagMatchRegexp('template'), (t) => replaceVueTemplateContent(ctx, t))
    .replace(buildTagMatchRegexp('script'), (t) => processJsContent(ctx, t))

  return result
}

function processJsContent(ctx: ProcessContext, fileContent: string): string {
  return fileContent
    .replace(templateStrRegexp, (text) => {
      if (!singleChineseRegexp.test(text)) {
        return text
      }

      const parsedResult = parseStringTemplate(text)

      if (!parsedResult.expressions.length) {
        const token = getI18nToken(ctx, text)
        return token || text
      }

      if (!singleChineseRegexp.test(parsedResult.tpl)) {
        return text.replace(tplInterpolationRegexp, (interopStr) => {
          const jsStr = interopStr.slice(2, -1)
          return '${' + processJsContent(ctx, jsStr) + '}'
        })
      }

      const variables = parsedResult.expressions.map((jsStr) => processJsContent(ctx, jsStr))

      const key = findTranslationKey(ctx, parsedResult.tpl)

      return key ? `$t('${key}', [${variables.join(', ')}])` : text;
    })
    .replace(quoteStrRegexp, (t) => {
      if (!singleChineseRegexp.test(t)) {
        return t
      }

      const token = getI18nToken(ctx, t)

      return token || t
    })
}

function replaceVueTemplateContent(ctx: ProcessContext, fileContent: string) {
  return (
    fileContent
      // replace props
      .replace(/(?<= ):?[\w\d_-]+="[^"]+"/g, (t) => {
        const eqlIdx = t.indexOf('=')
        let [key, value] = [t.slice(0, eqlIdx), t.slice(eqlIdx + 1)]

        if (!singleChineseRegexp.test(value)) {
          return t
        }

        const isSpecialExpressionKeys = key.startsWith('v-')
        const isExpression = key.startsWith(':') || isSpecialExpressionKeys

        const rawValue = value.slice(1, -1)

        if (isExpression) {
          const replacedText = processJsContent(ctx, rawValue)

          return replacedText === rawValue ? t : `${key}="${replacedText}"`
        } else {
          const token = getI18nToken(ctx, rawValue)

          return token ? `:${key}="${token}"` : t
        }
      })
      // replace {{ xxx }} js expression
      .replace(/{{(.|\n)+?}}/gm, (t) => {
        const replaced = processJsContent(ctx, t)

        return replaced
      })
      // replace normal text
      .replace(maybeChineseStrRegexp, (t) => {
        if (!singleChineseRegexp.test(t)) {
          return t
        }

        const tailingChars = /[:：,，\-、 ]+$/.exec(t)?.[0] || ''
        const leadingChars = /^[:：,，\-、 ]+/.exec(t)?.[0] || ''

        t = t.slice(leadingChars.length, tailingChars ? -tailingChars.length : undefined)

        const token = getI18nToken(ctx, t)

        return leadingChars + (token ? `{{ ${token} }}` : t) + tailingChars
      })
  )
}

function getI18nToken(ctx: ProcessContext, content: string) {
  const key = findTranslationKey(ctx, content)

  return key ? `$t('${key}')` : undefined
}

function findTranslationKey(ctx: ProcessContext, text: string) {
  if (/^('|"|\`)[\s\S]+?\1$/.test(text)) {
    text = text.slice(1, -1)
  }

  text = text.replace(/\n\s+/g, '\n')
  text = text.trim()

  const key = ctx.getI18nKey?.(text)

  return key
}
