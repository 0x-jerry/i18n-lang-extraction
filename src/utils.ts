export function buildTagMatchRegexp(tagName: string) {
  return new RegExp(`<${tagName}(.|\\n)*<\\/${tagName}>`, "g");
}

export function sortObjectKeys(obj: Record<string, any>) {
  const entries = Object.entries(obj);

  entries.sort((a, b) => a[0].localeCompare(b[0]));

  return Object.fromEntries(entries);
}

export function processWithoutComments(
  fileContent: string,
  fn: (s: string) => string
) {
  let i = 0;
  const comments: Record<string, string> = {};

  const updateCommentMap = (str: string, type: "html" | "js") => {
    const key = type === "html" ? `<!--${i++}-->` : `//__#${i++}__`;

    comments[key] = str;
    return key;
  };

  const content = fileContent
    .replace(htmlCommentRegexp, (s) => updateCommentMap(s, "html"))
    .replace(jsCommentRegexp, (s) => updateCommentMap(s, "js"));

  let result = fn(content);

  for (const [key, value] of Object.entries(comments)) {
    result = result.replace(key, () => value);
  }

  return result;
}

/**
 *
 * `xxx ${variable}` => { tpl: 'xxx {0}', expressions: ['variable']}
 * @param source
 * @returns
 */
export function parseStringTemplate(source: string) {
  const result = {
    tpl: "",
    expressions: [] as string[],
  };

  const tpl = magicReplace(source, tplInterpolationRegexp, (jsStr, idx) => {
    result.expressions.push(jsStr.slice(2, -1));
    return `{${idx}}`;
  });

  result.tpl = tpl === source ? source.slice(1, -1) : tpl;

  return result;
}

// ---------

const createQuoteInput = (char: string) =>
  `${char}(?:[^${char}\\\\]|\\\\.)*${char}`;

const doubleQuoteStr = createQuoteInput('"');
const singleQuoteStr = createQuoteInput("'");
const strTemplateQuoteStr = createQuoteInput("`");

const quoteStr = `(${doubleQuoteStr}|${singleQuoteStr})`;

/**
 * match single quote string or double quote string
 */
export const quoteStrRegexp = new RegExp(quoteStr, "g");

/**
 * match example:
 *
 * `xxx${xxx}bbb`
 * --------------
 */
export const templateStrRegexp = new RegExp(strTemplateQuoteStr, "g");

/**
 * match example:
 *
 * `aaa${xxx}bbb`
 *     -----
 */
export const tplInterpolationRegexp = /\$\{[\s\S]+?\}/g;

export const maybeChineseStrRegexp =
  /[\p{Script=Han}a-zA-Z0-9，,。\.?？!！"“”% \-、：:【】[\]]+/gmu;

export const singleChineseRegexp = /\p{Script=Han}/u;

export const jsCommentRegexp =
  /(?<=(?:\b|^|\s))(?:\/\/.+$|\/\*(?:(?:\s|\S)+?)\*\/)/gm;

export const htmlCommentRegexp = /<!--[\s\S]+?-->/g;

// -----------------

function magicSplit(source: string, pattern: RegExp) {
  const segments: { content: string; match?: RegExpMatchArray }[] = [];

  let lastMatch: RegExpMatchArray | undefined;

  for (const match of source.matchAll(pattern)) {
    const matchedStr = match[0];

    appendRawContent(match);

    segments.push({
      content: matchedStr,
      match,
    });

    lastMatch = match;
  }

  appendRawContent();

  return segments;

  function appendRawContent(match?: RegExpMatchArray) {
    const lastMatchIdx = lastMatch
      ? (lastMatch.index || 0) + lastMatch[0].length
      : 0;

    const nextMatchIndex = match ? match.index : source.length;

    if (lastMatchIdx !== nextMatchIndex) {
      const content = source.slice(lastMatchIdx, nextMatchIndex);

      segments.push({
        content,
      });
    }
  }
}

export function magicReplace(
  source: string,
  pattern: RegExp,
  replacer: (
    substring: string,
    index: number,
    match: RegExpMatchArray
  ) => string
) {
  const segments = magicSplit(source, pattern);

  let result = [];
  let idx = 0;
  for (const seg of segments) {
    if (seg.match) {
      result.push(replacer(seg.content, idx, seg.match));
      idx++;
    } else {
      result.push(seg.content);
    }
  }

  return result.join("");
}
