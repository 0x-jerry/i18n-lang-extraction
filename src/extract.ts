import { readFile } from 'fs/promises'
import { processI18nWorkflow } from './processWorkflow'

export async function extract(file: string) {
  const type = file.endsWith('.vue') ? 'vue' : 'js'
  const content = await readFile(file, 'utf-8')

  const result: string[] = []

  await processI18nWorkflow(content, {
    type,
    getI18nKey(text) {
      result.push(text)

      return '_#_'
    },
  })

  return result.sort()
}
