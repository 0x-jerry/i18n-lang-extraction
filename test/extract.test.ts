import fg from 'fast-glob'
import path from 'path'
import { extract } from '../src/extract'

describe('extract', async () => {
  const exampleDir = path.join(__dirname, 'examples')
  const files = await fg('**/*', {
    cwd: exampleDir,
  })

  for (const file of files) {
    it(`should work with: ${file}`, async () => {
      const filePath = path.join(exampleDir, file)
      const messages = await extract(filePath)

      expect(messages).toMatchSnapshot()
    })
  }
})
