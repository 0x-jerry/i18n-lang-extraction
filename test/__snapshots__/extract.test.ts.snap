// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extract > should work with: sample.ts 1`] = `
[
  "x1中文11",
  "中{0}文{1}",
  "中文",
]
`;

exports[`extract > should work with: sample.tsx 1`] = `
[
  "x1中文11",
  "中{0}文{1}",
  "中文",
]
`;

exports[`extract > should work with: sample.vue 1`] = `
[
  "x1中文11",
  "不错12312哟",
  "不错哟",
  "中{0}文",
  "中{0}文{1}",
  "中文",
  "这是一段文本",
]
`;
